api/v1alpha1/author_webhook.go

```diff
@@ -0,0 +1,197 @@
+func (a *authorWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	switch req.Kind.Kind {
+	case "Release":
+		return a.handleRelease(req)
+	case "ReleasePlan":
+		return a.handleReleasePlan(req)
+	default:
+		return admission.Errored(http.StatusInternalServerError, errors.New("webhook tried to handle an unsupported resource: "+string(req.Kind.Kind)))
+	}
+}
+
+// handleRelease takes an incoming admission request and returns an admission response. Create requests
+// add an author label with the current user. Update requests are rejected if the spec or author label
+// are being modified. All other requests are accepted without action.
+func (a *authorWebhook) handleRelease(req admission.Request) admission.Response {
+	var releaseNew Release
+	err := json.Unmarshal(req.Object.Raw, &releaseNew)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, errors.Wrap(err, "error decoding object"))
+	}
+	switch req.AdmissionRequest.Operation {
+	case admissionv1.Create:
+		return a.patchAuthorResponse(req.UserInfo.Username, &releaseNew)
+	case admissionv1.Update:
+		var releaseOld Release
+		err = json.Unmarshal(req.OldObject.Raw, &releaseOld)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, errors.Wrap(err, "error decoding object"))
+		}
+		if releaseNew.GetLabels()[metadata.AuthorLabel] != releaseOld.GetLabels()[metadata.AuthorLabel] {
+			return admission.Errored(http.StatusInternalServerError, errors.New("release author label cannnot be updated"))
+		}
+	}
+	return admission.Allowed("Success")
+}
+
+// handleReleasePlan takes an incoming admission request and returns an admission response. The request
+// is rejected or denied based on the value of the attribution label, with the author label being added or
+// removed. The attribution label should always exist and be set to true if the author label is set.
+func (a *authorWebhook) handleReleasePlan(req admission.Request) admission.Response {
+	var releasePlanNew ReleasePlan
+	err := json.Unmarshal(req.Object.Raw, &releasePlanNew)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, errors.Wrap(err, "error decoding object"))
+	}
+	switch req.AdmissionRequest.Operation {
+	case admissionv1.Create:
+		if releasePlanNew.GetLabels()[metadata.AttributionLabel] == "true" {
+			author := releasePlanNew.GetLabels()[metadata.AuthorLabel]
+			if author != "" && author != authorLabelValue(req.UserInfo.Username) {
+				return admission.Errored(http.StatusInternalServerError, errors.New("[labels] cannot set attribution to true if not setting"+
+					" current user as author. Expected author is: "+authorLabelValue(req.UserInfo.Username)))
+			}
+			return a.patchAuthorResponse(req.UserInfo.Username, &releasePlanNew)
+		}
+		return admission.Allowed("Success")
+	case admissionv1.Update:
+		var releasePlanOld ReleasePlan
+		err = json.Unmarshal(req.OldObject.Raw, &releasePlanOld)
```

Comment:
This is exactly the same as the one at the beginning of the function.

Reply:
No it is not. One is the Object and one is the OldObject